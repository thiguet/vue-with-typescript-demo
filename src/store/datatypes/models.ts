interface Unique {
    id?: string;
}

export enum Measures {
    UN = 'UN',
    L = 'L',
    KG = 'KG',
    G = 'G',
}

export interface Product extends Unique {
    name: string;
    measure: Measures;
    qtd: number;
    minQtd: number;
    image?: string;
}

export interface User extends Unique {
    name: string;
    email: string;
}

export interface ReportItem {
    text: string;
    value: number;
}

export type ReportData = ReportItem[];

export interface QuantityReportData {
    productName: string;
    productQuantity: number;
}

export type QuantityReport = QuantityReportData[];
export interface MeasuresReportData {
    measureName: string;
    measureQuantity: number;
}

export type MeasuresReport = MeasuresReportData[];

export enum VuexAppModules {
    products = 'products',
    login = 'login',
    alert = 'alert',
    reports = 'reports',
}
