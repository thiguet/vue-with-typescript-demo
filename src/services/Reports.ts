import { QuantityReport, MeasuresReport } from '@/store/datatypes/models';

import VueWithTSAPI from '@/services/';

export { VueWithTSAPI };

export async function getQuantityReport(): Promise<QuantityReport> {
    return (await VueWithTSAPI.get('/reports/quantity')).data as QuantityReport;
}

export async function getMeasuresReport(): Promise<MeasuresReport> {
    return (await VueWithTSAPI.get('/reports/measures')).data as MeasuresReport;
}
