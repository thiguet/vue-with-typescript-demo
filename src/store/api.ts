import axios from 'axios';
import { LoginSubmit, User } from './datatypes/models';

export const conduitAPI = axios.create({
    baseURL: 'https://conduit.productionready.io/api',
});

export function setJWT(jwt: string) {
    conduitAPI.defaults.headers.common.Authorization = `Token ${jwt}`;
}

export function clearJWT() {
    delete conduitAPI.defaults.headers.common.Authorization;
}

export async function login(user: LoginSubmit): Promise<User> {
    const response = await axios.post('/users/login', {
        user,
    });
    return response.data as User;
}
