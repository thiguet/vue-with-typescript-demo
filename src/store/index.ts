import Vue from 'vue';
import Vuex from 'vuex';
import { createStore, Module } from 'vuex-smart-module';

import LoginModule from '@/store/modules/login';
import ProductsModule from '@/store/modules/products';
import AlertModule from '@/store/modules/alert';
import { VuexAppModules } from './datatypes/models';

Vue.use(Vuex);

const root = new Module({
    modules: {
        [VuexAppModules.products]: ProductsModule,
        [VuexAppModules.login]: LoginModule,
        [VuexAppModules.alert]: AlertModule,
    },
});

export default createStore(
    root,
    {
        strict: process.env.NODE_ENV !== 'production',
    },
);
