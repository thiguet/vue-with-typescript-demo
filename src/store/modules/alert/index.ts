import { Module } from 'vuex-smart-module';
import Mutations, { MutationTypes as MTypes } from './mutations';

export const MutationTypes = MTypes;
export const AlertMutations = Mutations;

export class AlertState {
  message!: string;

  display!: boolean;
}

export default new Module({
    state: AlertState,
    mutations: AlertMutations,
});
