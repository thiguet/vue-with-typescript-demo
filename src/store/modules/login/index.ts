import { Module } from 'vuex-smart-module';
import Mutations, { MutationTypes } from './mutations';
import Actions, { ActionTypes } from './actions';
import State from './state';

export { State, Mutations, MutationTypes, Actions, ActionTypes };

export default new Module({
    state: State,
    mutations: Mutations,
    actions: Actions,
});
