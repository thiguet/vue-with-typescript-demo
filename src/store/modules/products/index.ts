import { Module } from 'vuex-smart-module';
import State from './state';
import ProductsMutations, { MutationTypes as MTypes } from './mutations';
import ProductsActions, { ActionTypes as ATypes } from './actions';

export type ProductsState = State;
export const MutationTypes = MTypes;
export const ActionTypes = ATypes;

export default new Module({
  state: State,
  mutations: ProductsMutations,
  actions: ProductsActions,
});
