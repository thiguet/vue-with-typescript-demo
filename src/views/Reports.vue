<template>
    <div id="reports">
        <div class="charts-wrapper">
            <ScalerBox v-if="!isLoading" box-title="Produtos por Peso %">
                <DonutChart :data="donutChartData" />
            </ScalerBox>

            <ScalerBox v-if="!isLoading" box-title="Produtos em Estoque">
                <LineChart :data="lineChartData" />
            </ScalerBox>
        </div>

        <Button
            id="go-back"
            name="voltar"
            label="Voltar"
            icon="./assets/icons/goback.svg"
            class="footer-btn goback"
            :onclick="routeToHomePage"
        />
    </div>
</template>

<script lang="ts">
import Vue from 'vue';
import Component from 'vue-class-component';

import { namespace } from 'vuex-class';

import { ActionTypes, Actions } from '@/store/modules/reports';

import {
    MeasuresReport,
    QuantityReport,
    ReportData,
    VuexAppModules,
} from '@/store/datatypes/models';

import DonutChart from '@/components/DonutChart.vue';
import LineChart from '@/components/LineChart.vue';
import ScalerBox from '@/components/ScalerBox.vue';
import Button from '@/components/Button.vue';
import { ReportsView } from './models.d';

const reports = namespace(VuexAppModules.reports);

const { Action, State } = reports;

@Component({
    components: {
        Button,
        DonutChart,
        LineChart,
        ScalerBox,
    },
})
export default class Reports extends Vue implements ReportsView {
    @State
    private isLoading!: boolean;

    @State
    private measuresReport!: MeasuresReport;

    @State
    private quantityReport!: QuantityReport;

    @Action
    private fetchMeasuresReport!: Actions[ActionTypes.fetchMeasuresReport];

    @Action
    private fetchQuantityReport!: Actions[ActionTypes.fetchQuantityReport];

    created() {
        this.fetchReportData();
    }

    get donutChartData(): ReportData {
        return this.measuresReport.map((item) => ({
            text: item.measureName,
            value: +item.measureQuantity,
        }));
    }

    get lineChartData(): ReportData {
        return this.quantityReport.map((item) => ({
            text: item.productName,
            value: +item.productQuantity,
        }));
    }

    private fetchReportData() {
        this.fetchMeasuresReport();
        this.fetchQuantityReport();
    }

    routeToHomePage() {
        this.$router.push('/');
    }
}
</script>

<style scoped>
#reports {
    flex-flow: column;
}

.charts-wrapper {
    align-items: center;
    height: 85vh;
    flex-flow: row;
    justify-content: space-around;
}

.goback button {
    height: 60px;
    width: 150px;
    justify-content: center;
    display: flex;
}

@media screen and (max-width: 600px) {
    .charts-wrapper {
        flex-flow: column;
    }
}
</style>
