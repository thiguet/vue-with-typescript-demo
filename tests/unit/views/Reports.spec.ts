import { shallowMount, mount, createLocalVue, Wrapper } from '@vue/test-utils';

import Vuex, { ModuleTree, Store } from 'vuex';
import VueRouter from 'vue-router';

import Reports from '@/views/Reports.vue';
import {
    MeasuresReport,
    QuantityReport,
    VuexAppModules,
} from '@/store/datatypes/models';
import { State, MutationTypes, ActionTypes } from '@/store/modules/reports';

import { ReportsVuex } from '../store/models';
import {
    getFakeMeasuresReportData,
    getFakeQuantityReportData,
} from '../utils/ReportsFactory';

jest.mock('@/services/Reports');

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(VueRouter);

const router = new VueRouter();

describe('Reports', () => {
    let store: Store<State>;

    let reports: ReportsVuex;

    let quantityReport: QuantityReport;
    let measuresReport: MeasuresReport;

    const build = () => {
        reports = {
            namespaced: true,
            state: {
                isLoading: false,
                quantityReport,
                measuresReport,
            },
            getters: {},
            mutations: {
                [MutationTypes.setMeasuresReport]: jest.fn(),
                [MutationTypes.setQuantityReport]: jest.fn(),
            },
            actions: {
                [ActionTypes.fetchMeasuresReport]: jest.fn(),
                [ActionTypes.fetchQuantityReport]: jest.fn(),
            },
        };

        const modules: ModuleTree<State> = {
            [VuexAppModules.reports]: {
                ...reports,
            },
        };

        store = new Vuex.Store({
            modules,
        });

        const options = {
            localVue,
            store,
            router,
        };

        const wrapper: Wrapper<Reports> = shallowMount(Reports, options);
        const mountedWrapper: Wrapper<Reports> = mount(Reports, options);

        return {
            wrapper,
            donutChart: () => mountedWrapper.find('#donut-chart'),
            lineChart: () => mountedWrapper.find('#line-chart'),
            goBackButton: () => mountedWrapper.find('#go-back'),
        };
    };

    beforeEach(() => {
        quantityReport = getFakeQuantityReportData();
        measuresReport = getFakeMeasuresReportData();

        router.push = jest.fn();
    });

    it('renders component', () => {
        quantityReport = [
            {
                productName: 'teste',
                productQuantity: 10,
            },
        ];
        measuresReport = [
            {
                measureName: 'teste',
                measureQuantity: 20,
            },
        ];

        const { wrapper } = build();
        expect(wrapper).toMatchSnapshot();
    });

    it('renders main components', () => {
        const { lineChart, donutChart } = build();
        expect(donutChart().exists()).toBe(true);
        expect(lineChart().exists()).toBe(true);
    });

    it('go back to home page', async () => {
        const { goBackButton } = build();

        await goBackButton().trigger('click');

        expect(router.push).toHaveBeenCalledWith('/');
    });

    it('must fetch data on mounting', () => {
        const { actions } = reports;

        build();

        expect(actions.fetchQuantityReport).toBeCalled();
        expect(actions.fetchMeasuresReport).toBeCalled();
    });

    it('donut chart must receive vuex bindings', async () => {
        const { donutChart } = build();
        expect(
            donutChart()
                .props()
                .data.sort((a: any, b: any) => a.value - b.value),
        ).toEqual(
            measuresReport
                .map((MRItem) => ({
                    text: MRItem.measureName,
                    value: MRItem.measureQuantity,
                }))
                .sort((a, b) => a.value - b.value),
        );
    });

    it('line chart must receive vuex bindings', async () => {
        const { lineChart } = build();
        expect(
            lineChart()
                .props()
                .data.sort((a: any, b: any) => a.value - b.value),
        ).toEqual(
            quantityReport
                .map((MRItem) => ({
                    text: MRItem.productName,
                    value: MRItem.productQuantity,
                }))
                .sort((a, b) => a.value - b.value),
        );
    });
});
