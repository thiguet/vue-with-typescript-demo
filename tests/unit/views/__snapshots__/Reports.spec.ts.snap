// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reports renders component 1`] = `
<div id="reports">
  <div class="charts-wrapper">
    <scalerbox-stub boxtitle="Produtos por Peso %">
      <donutchart-stub data="[object Object]"></donutchart-stub>
    </scalerbox-stub>
    <scalerbox-stub boxtitle="Produtos em Estoque">
      <linechart-stub data="[object Object]"></linechart-stub>
    </scalerbox-stub>
  </div>
  <button-stub id="go-back" name="voltar" onclick="function () { [native code] }" label="Voltar" icon="./assets/icons/goback.svg" class="footer-btn goback"></button-stub>
</div>
`;
